daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala

```diff
@@ -5,22 +5,105 @@ package com.daml.lf
-    Utf8.Ordering.compare(text1, text2)
+  @throws[SError.SErrorCrash]
+  // Ordering between two SValues of same type.
+  // This follows the equality defined in the daml-lf spec.
+  def compare(x: SValue, y: SValue): Int = {
+    import SValue._
 
-  private def compareCid(cid1: ContractId, cid2: ContractId): Int =
+    var diff = 0
+    var stackX = List(Iterator.single(x))
+    var stackY = List(Iterator.single(y))
+    // invariant: stackX.length == stackY.length
+
+    @inline
+    def push(xs: Iterator[SValue], ys: Iterator[SValue]): Unit = {
+      stackX = xs :: stackX
+      stackY = ys :: stackY
+    }
+
+    @inline
+    def pop(): Unit = {
+      stackX = stackX.tail
+      stackY = stackY.tail
+    }
+
+    @inline
+    def step(tuple: (SValue, SValue)) =
+      tuple match {
+        case (SUnit, SUnit) =>
+        case (SBool(x), SBool(y)) =>
+          diff = x compareTo y
+        case (SInt64(x), SInt64(y)) =>
+          diff = x compareTo y
+        case (SNumeric(x), SNumeric(y)) =>
+          diff = x compareTo y
+        case (SText(x), SText(y)) =>
+          diff = Utf8.Ordering.compare(x, y)
+        case (SDate(x), SDate(y)) =>
+          diff = x compareTo y
+        case (STimestamp(x), STimestamp(y)) =>
+          diff = x compareTo y
+        case (SParty(x), SParty(y)) =>
+          // party are ASCII, so UTF16 comparison matches UTF8 comparison.
+          diff = x compareTo y
+        case (SContractId(x), SContractId(y)) =>
+          diff = compareCid(x, y)
+        case (SEnum(_, _, xRank), SEnum(_, _, yRank)) =>
+          diff = xRank compareTo yRank
+        case (SRecord(_, _, xs), SRecord(_, _, ys)) =>
+          push(xs.iterator().asScala, ys.iterator().asScala)
+        case (SVariant(_, _, xRank, x), SVariant(_, _, yRank, y)) =>
+          push(Iterator.single(x), Iterator.single(y))
+          diff = xRank compareTo yRank
+        case (SList(xs), SList(ys)) =>
+          push(xs.iterator, ys.iterator)
+        case (SOptional(xOpt), SOptional(yOpt)) =>
+          push(xOpt.iterator, yOpt.iterator)
+        case (STextMap(xMap), STextMap(yMap)) =>
+          val xKeys = xMap.keys.toSeq.sorted
+          val yKeys = yMap.keys.toSeq.sorted
+          push(
+            new InterlacedIterator(xKeys.iterator.map(SText), xKeys.iterator.map(xMap)),
+            new InterlacedIterator(yKeys.iterator.map(SText), yKeys.iterator.map(yMap)),
+          )
+        case (SGenMap(xMap), SGenMap(yMap)) =>
+          push(
+            new InterlacedIterator(xMap.keys.iterator, xMap.values.iterator),
+            new InterlacedIterator(yMap.keys.iterator, yMap.values.iterator),
+          )
+        case (SStruct(_, xs), SStruct(_, ys)) =>
+          push(xs.iterator().asScala, ys.iterator().asScala)
+        case (SAny(xType, x), SAny(yType, y)) =>
+          push(Iterator.single(x), Iterator.single(y))
+          diff = TypeOrdering.compare(xType, yType)
```

Comment:
Just to reflect the actual order of comparison in code.
```suggestion
          diff = TypeOrdering.compare(xType, yType)
          push(Iterator.single(x), Iterator.single(y))
```